### ISSUE TITLE
Allow template local variables to publish type information so that IDEs can provide better completion

### ISSUE THREAD

**Issue by mhevery** [angular/angular#4850]
## Problem

Assume the following code example:

```
@Component({
  template: `
  <ul>
    <li *ngFor="let user of users; let i = index">{{i + 0}}: {{user.last}}, {{user.first}}<li>
  <ul>
`
})
class MyExample {
  users: List<User> = [];
}

class User {
  first: string;
  last: string;
}
```

Here the IDE has type information for `users`, but it does not have type information for `user` and `i`. For this reason it can not verify:
 - that `i+0` is valid because it does not know that `i` is a number.
 - that `user.last` and `user.first` are valid because it does not know what is the type for `user`

## Solution

Parameterize `TemplateRef<T>`.

```
class Row {
  $implicit: any;
  index: number;
  even: boolean;
  odd: boolean;
}

@Directive({
  selector: '[ng-for][ng-for-of]',
})
class NgFor  {
  constructor(templateRef: TemplateRef<Row>, viewContainer:ViewContainer) {
  }  
}

```

## Changing the View API

Currently setting the local variables is done as:

```
var view: ViewRef;
view.setLocal('$implicit', item)
```

A better way to do this, which is more consistent with the above proposal is to inject a locals object during view Creation.

```
var local: Row = new Row(item, index);
var view: ViewRef<Row> = this._viewContainer.createEmbeddedView(templateRef, index, local);
```

The advantage is that currently `ngFor` has to compute `odd`, `even` regardless of wether it is needed in the template. By injecting a local's object the `odd` field can be computed.

```
class Row implements Locals {
  constructor(public $implicit: T, index: number) {}

  get even() => this.index % 2 == 0;
  get odd() => this.index % 2 == 1;
}
```

This means that the `Row` object can have a rich API, but unless a field is actually referenced from the template it has no cost of computing these fields. Furthermore by using a typed object we can get better performance on locals then when using hashes. 

## Add Type to the templates

Requires: read generics from DI dependencies (constructor arguments).

```typescript
import {User} from '...';

@Component({
  types: {User}
  template: `
  <ul>
    <li *ngFor="let user: User of users; let i = index">{{i + 0}}: {{user.last}}, {{user.first}}<li>
  <ul>
`
})
class MyExample {
  users: List<User> = [];
}
```

Notice the `types` in the `@Component` which sets up a map of template types to real types. This will allow the compiler to generate code such as this.

```
Generated code:

// Deduce generic Row from NgFor constructor asking for a TemplateRef<Row>
class MyExample_View1<Row> extends AppView<Row> {
  ...
  changeDetection() {
    var user: User /* retrieved from the Template */ = context.$implicit;
    var i = context.index; // Here i is number because context.index is number.
  }

}

class AppView<T> {
  context: T
}

class TemplateRef<T> {
  appView: AppView<T>;
  get context(): T {
    return this.appView.context;
  }
}
```

**Comment by scheglov** [comment]
There are couple of things I don't understand about the proposed solution.

1. Is every directive declaring local variables going to provide `LocalesFactory` with a type argument, like `Row` in case of `ng-for`?

2. What is `@Input() in: Array<T>;`? Currently `NgFor` declares the `ng-for-of` property. I expected to see it instead of `in`.

3. Is the expectation that IDE will "solve" the set of "type equations" against type parameters using properties in order to know exact types of the `Locals` subclass fields?

**Comment by mhevery** [comment]
1. Yes, a directive would have to implement `LocalesFactory` to give IDE hints.
2. Sorry that is a typo, changed to `of`.
3. Yes the IDE will have to solve the type equations to propagate the types. Is that a problem? If you have ideas on how to make that better/simpler please share.

**Comment by scheglov** [comment]
No, it's not a problem.
I just want to make sure that I correctly understand the solution and how you see it to be used.

Thank you.
I think this will give the Angular2 analysis plugin enough type information.

**Comment by tbosch** [comment]
@scheglov Could you have another look? We just simplified the proposal to not rely on a new interface `LocalsFactory` but rather to make `ViewContainer` be parameterized on the locals type. Would this also work?

**Comment by scheglov** [comment]
Yes, this works too.

**Comment by mhevery** [comment]
@scheglov There is a bit of a change of how we will implement. Please have a look again.

**Comment by scheglov** [comment]
Works for me.
This is even easier for static analysis.

**Comment by tbosch** [comment]
Quick update:
- the first part is done, i.e. `TemplateRef` now takes a type as argument.
- still missing: `types` property on the `@Component` annotation.

**Comment by tbosch** [comment]
Also, once we support the `types` property on `@Component`, we could also use this for supporting enums and static properties in the template in expressions.

**Comment by tbosch** [comment]
The missing part can be done post final as it is not a breaking change.

**Comment by mhevery** [comment]
I think https://github.com/angular/angular/issues/10510 is a better solution to this problem.

**Comment by chuckjaz** [comment]
Closed in favor of #12398
